### PR TITLE
Fix directory permissions for multi-directory globs. Follow-up #50559

### PR DESCRIPTION
**Pull request by zvonand**
Follow-up for #50559.

Add setting `ignore_access_denied_multidirectory_globs` to avoid `permission denied` in case there are inaccessible directories/files.

### Explanation
Having the following structure in `user_files`:
```
user_files/
├── data1
│   ├── f1.csv
├── data2
│   ├── f2.csv
└── test_root
```

Where `data1`, `data2` are accessible by CH, but no rights to read `test_root`. 
CH would throw:
`Code: 1001. DB::Exception: std::__1::__fs::filesystem::filesystem_error: filesystem error: in directory_iterator::directory_iterator(...): Permission denied`
for a query like
`SELECT *, _path, _file FROM file('{data1/f1,data2/f2}.csv', CSV)`.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add setting `ignore_access_denied_multidirectory_globs`.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit e851be78bce89b5cf652c710714277824c4f85e8 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52839/e851be78bce89b5cf652c710714277824c4f85e8/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by zvonand** [comment]
**PullRequestCI / SQLLogicTestRelease**
Looks unrelated (and occasional), the same seen [here](https://github.com/ClickHouse/ClickHouse/pull/51985#issuecomment-1627782631). Looks like some race is going on.

```+ /clickhouse-tests/sqllogic/runner.py --log-file /test_output/runner-statements-test.log --log-level info statements-test --input-dir /sqllogictest --out-dir /test_output/statements-test
+ ts '%Y-%m-%d %H:%M:%S'
2023-08-01 22:29:16 multiprocessing.pool.RemoteTraceback: 
2023-08-01 22:29:16 """
2023-08-01 22:29:16 Traceback (most recent call last):
2023-08-01 22:29:16   File "/usr/lib/python3.10/multiprocessing/pool.py", line 125, in worker
2023-08-01 22:29:16     result = (True, func(*args, **kwds))
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/runner.py", line 135, in _child_process
2023-08-01 22:29:16     runner.run_all_tests_from_file(test, input_dir)
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/test_runner.py", line 557, in run_all_tests_from_file
2023-08-01 22:29:16     self.run_one_test(stream, test_name, test_file)
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/test_runner.py", line 532, in run_one_test
2023-08-01 22:29:16     with self.connection.with_one_test_scope():
2023-08-01 22:29:16   File "/usr/lib/python3.10/contextlib.py", line 142, in __exit__
2023-08-01 22:29:16     next(self.gen)
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/connection.py", line 164, in with_one_test_scope
2023-08-01 22:29:16     self.drop_all_tables()
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/connection.py", line 136, in drop_all_tables
2023-08-01 22:29:16     logger.info("tables will be dropped: %s", list_result.get_result())
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/connection.py", line 241, in get_result
2023-08-01 22:29:16     self.assert_no_exception()
2023-08-01 22:29:16   File "/clickhouse-tests/sqllogic/connection.py", line 264, in assert_no_exception
2023-08-01 22:29:16     raise ProgramError(
2023-08-01 22:29:16 exceptions.ProgramError: request doesn't have a result set, it has the exception, exception: ('HY000', '[HY000] HTTP status code: 500\nReceived error:\nCode: 373. DB::Exception: Session clickhouse_odbc_4113627657914525518 is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.8.1.327)\n\n (1) (SQLExecDirectW)')
2023-08-01 22:29:16 """
```

File: src/Storages/StorageFile.cpp [hunk 1]
```diff
@@ -121,7 +121,7 @@ void listFilesWithFoldedRegexpMatchingImpl(const std::string & path_for_ls,
         return;
 
     const fs::directory_iterator end;
-    for (fs::directory_iterator it(path_for_ls); it != end; ++it)
+    for (fs::directory_iterator it(path_for_ls, fs::directory_options::skip_permission_denied); it != end; ++it)
```
Comment (SmitaRKulkarni) [inline src/Storages/StorageFile.cpp:124]: @zvonand : In the example, `SELECT *, _path, _file FROM file('{data1/f1,data2/f2}.csv', CSV)` if user doesn't have rights to `data2` looks like we would just skip it without throwing any error.

Comment (zvonand) [reply src/Storages/StorageFile.cpp:124]: @SmitaRKulkarni 
IMO we could append the docs with something like `a glob with '/' in curly braces will ignore directories that cannot be accessed`. 
I don't think that making the code twice as unreadable is a good idea for such a small issue.
Also, it doesn't affect _old_ workflow, it is _only_ about `{foo/bar,x/y/z}` globs.

**Comment by alexey-milovidov** [comment]
Throwing an exception is better than silently ignoring the files.
You can introduce a new behavior under a setting, but it should be disabled by default.

**Comment by zvonand** [comment]
> Throwing an exception is better than silently ignoring the files.

In general, sure it is. But this limits the use cases, as there are often inaccessible directories where you don't need to go :). Maybe

Also, I added a docs update to this PR. It does not affect behavior prior to #50559. So I don't think it is a big deal to modify a thing that no one has used yet.

However, a setting may be a good idea.

File: src/Storages/StorageFile.cpp [hunk 2]
```diff
@@ -203,7 +210,7 @@ void listFilesWithRegexpMatchingImpl(
     re2::RE2 matcher(regexp);
     if (!matcher.ok())
         throw Exception(ErrorCodes::CANNOT_COMPILE_REGEXP,
-            "Cannot compile regex from glob ({}): {}", for_match, matcher.error());
+            "Cannot compile regex from gllistFilesWithFoldedRegexpMatchingImplob ({}): {}", for_match, matcher.error());
```
Comment (SmitaRKulkarni) [inline src/Storages/StorageFile.cpp:213]:
```suggestion
            "Cannot compile regex from glob ({}): {}", for_match, matcher.error());
```

File: docs/en/operations/settings/settings.md [hunk 3]
```diff
@@ -4614,3 +4614,43 @@ SELECT toFloat64('1.7091'), toFloat64('1.5008753E7') SETTINGS precise_float_pars
...
+└── test_root
+```
+where `data1`, `data2` directories are accessible, but one has no rights to read `test_root` directories.
+
+For a query like `SELECT *, _path, _file FROM file('my_directory/{data1/f1,data2/f2}.csv', CSV)` an exception will be thrown:
+`Code: 1001. DB::Exception: std::__1::__fs::filesystem::filesystem_error: filesystem error: in directory_iterator::directory_iterator(...): Permission denied`.  
+It happens because a multi-directory glob requires a recursive search in _all_ available directories under `my_directory`.
+
+If this setting is on, all inaccessible directories will be silently skipped, even if they are explicitly specified inside `{}`.
+
```
Comment (SmitaRKulkarni) [inline docs/en/operations/settings/settings.md:4644]: This setting is only applicable when multi directory glob is used.

File: tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+${CLICKHOUSE_CLIENT} --query "SELECT *, _file FROM file('${user_files_path}/${CLICKHOUSE_TEST_UNIQUE_NAME}/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data1.csv';"
+${CLICKHOUSE_CLIENT} --query "SELECT *, _file FROM file('${user_files_path}/${CLICKHOUSE_TEST_UNIQUE_NAME}/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data2.csv';"
+
+# Add a directory to test against permission_denied
+rm -rf ${user_files_path:?}/${CLICKHOUSE_TEST_UNIQUE_NAME:?}/dir_inaccessible/ && mkdir -p ${user_files_path}/${CLICKHOUSE_TEST_UNIQUE_NAME}/dir_inaccessible/ && chmod 000 ${user_files_path}/${CLICKHOUSE_TEST_UNIQUE_NAME}/dir_inaccessible/
+
+${CLICKHOUSE_CLIENT} --query "SELECT *, _file FROM file('${user_files_path}/${CLICKHOUSE_TEST_UNIQUE_NAME}/dir{?/subdir?1/da,2/subdir2?/da}ta1.csv', CSV) SETTINGS ignore_eacces_multidirectory_globs = 1;"
+
+rm -rf ${user_files_path:?}/${CLICKHOUSE_TEST_UNIQUE_NAME:?}
```
[the remark this answers is not in the document]
Comment (zvonand) [reply tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh:41]: > Maybe we can add a test for normal glob...

I cannot easily add a stateless test for this, I started but abandoned it. Stateless can be run in 2 different ways: locally and in docker. 
In docker, everything is run as root, thus I cannot make something unreadable, I will need to create/delete a user in a test. If I then try to run the same test locally -- it will need root access (which is totally crazy for a test)

[the remark this answers is not in the document]
Comment (zvonand) [reply tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh:41]: TBH I don't even know where to put this test -- it doesn't fit stateless, it is not integration (it doesn't 'integrate' with anything).

The problem here is that as a root (as Docker runs) I have read access to everything. Even if there is someone else's directory with 000 permissions, I can read it as root. So, this doesn't go to stateless.

File: tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+# Tags: no-replicated-database, no-parallel
+
+#SELECT *, _file FROM file('02771/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data1.csv';
+#SELECT *, _file FROM file('02771/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data2.csv';
```
Comment (SmitaRKulkarni) [inline tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh:5]: commented lines

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: docs/en/operations/settings/settings.md [hunk 6]
```diff
@@ -4643,3 +4643,44 @@ SELECT toFloat64('1.7091'), toFloat64('1.5008753E7') SETTINGS precise_float_pars
...
+Allows to ignore 'permission denied' errors when using multi-directory `{}` globs for [File](../../sql-reference/table-functions/file.md#globs_in_path) and [HDFS](../../sql-reference/table-functions/hdfs.md) storages.
+This setting is only applicable when multi directory `{}` glob is used.
+
+Possible values: `0`, `1`.
+
+Default value: `0`.
+
+### Example
+
+Having the following structure in `user_files`:
```
[the remark this answers is not in the document]
Comment (zvonand) [reply docs/en/operations/settings/settings.md:4658]: well, I meant the default directory where CH would put files. It's called `user_files` by default

File: tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# Tags: no-replicated-database, no-parallel
+
+set -eu
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# Data preparation.
+# Now we can get the user_files_path by use the table file function for trick. also we can get it by query as:
```
Comment (SmitaRKulkarni) [inline tests/queries/0_stateless/02771_multidirectory_globs_storage_file.sh:11]:
```suggestion
# Now we can get the user_files_path by using the file function, we can also get it by this query:
```
I know we have this comment in few other tests, but could you please fix this typo here.

**Comment by zvonand** [comment]
@SmitaRKulkarni updated according to your comments, all checks passed.

File: docs/en/operations/settings/settings.md [hunk 8]
```diff
@@ -4659,3 +4659,44 @@ The default value is `false`.
 ``` xml
 <validate_tcp_client_information>true</validate_tcp_client_information>
 ```
+
+## ignore_eacces_multidirectory_globs {#ignore_eacces_multidirectory_globs}
```
Comment (alexey-milovidov) [inline docs/en/operations/settings/settings.md:4663]: 1. Typo: `acces` -> `access`.

2. `eaccess` sounds weird.
I know that it is from Linux's/POSIX's `EACCESS`.
But let's not follow this notation.

`ignore_paths_without_permission_in_multidirectory_globs` will be better.

> This setting is only applicable when multi directory `{}` glob is used.

This is difficult to understand.
I didn't read the changes, didn't read the PR description, but from the docs, it makes me wonder:
- why doesn't it apply for `**` which is also a multi-directory glob;
- why don't we have a similar setting to skip files (not directories) without permissions when using simple globs?

Comment (zvonand) [reply docs/en/operations/settings/settings.md:4663]: > why doesn't it apply for ** which is also a multi-directory glob    

`**` skips them anyway, isn't it?

> why don't we have a similar setting to skip files (not directories) without permissions when using simple globs?

Well, I was just addressing the problem that came with newly introduced feature. Previously, globs worked as is and everyone seemed to be satisfied.
DYM this setting shall be applied to all globs? Or you mean that it shall be explicitly specified in docs that it doesn't work for other settings?

Comment (zvonand) [reply docs/en/operations/settings/settings.md:4663]: Minor:

> Typo: acces -> access

No, not a typo, it is `eacces`

> But let's not follow this notation.

Why? It doesn't really matter to me personally 😄 But `eacces` is a rather conventional name, and it would be easy to find if you know what you're looking for. I don't think `ignore_paths_without_permission_in_multidirectory_globs` is really better -- isn't it a bit too long?
There are also references to this setting from `globs` sections, so it still can be found.

Comment (zvonand) [reply docs/en/operations/settings/settings.md]: Anyway, setting renamed

**Comment by alexey-milovidov** [comment]
In the example from the description:

> Where data1, data2 are accessible by CH, but no rights to read test_root.
> SELECT *, _path, _file FROM file('{data1/f1,data2/f2}.csv', CSV).

Why this query will even attempt to access test_root?

**Comment by zvonand** [comment]
> Why this query will even attempt to access test_root?

See [this comment](https://github.com/ClickHouse/ClickHouse/pull/52839#discussion_r1284555330)

tldr: because there is no way to match a regex prefix against a string (kind of "reverse" matching -- match a prefix of a regex against a given string)  
Trying to do this means reinventing regex (constructing a fsm that parses regex manually, and that will definitely be big, non-optimal and error-prone for sure)

**Comment by alexey-milovidov** [comment]
Ok. So, we need to traverse the whole directory tree to match by paths (rather than by separate components).

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
But it is still a problem - we are doing it the wrong way.

`{x,y}` - this is a generator, not a matcher.
It should iteratively generate `x` and `y` instead of constructing a regexp and matching by regexp.

**Comment by alexey-milovidov** [comment]
We will have to rewrite all of this in the future and drop the setting you've introduced :(

**Comment by zvonand** [comment]
> `{x,y}` - this is a generator, not a matcher. It should iteratively generate `x` and `y` instead of constructing a regexp and matching by regexp.

Didn't get it -- what does this mean? User-facing behavior shall be the same, right?

**Comment by zvonand** [comment]
> We will have to rewrite all of this in the future and drop the setting you've introduced :(

I will do it in a while.

**Comment by hodgesrm** [comment]
Hi @zvonand, 

What is the proposed fix? 
Seems as if you could parse the string, conclude there are no globs, and generate the list. This would eliminate the need to search (performance hit) and eliminate permission issues.

**Comment by zvonand** [comment]
@alexey-milovidov could you please put a `can be tested` to #54863 (follow-up for these two PRs) ?  
Let me finish this and forget about it 😃

**Comment by Blargian** [comment]
Setting `ignore_access_denied_multidirectory_globs` seems to be missing from settings.h. We reference it [here](https://github.com/ClickHouse/ClickHouse/blob/e184fac48933fab21e141bca91c4bbc27264e599/docs/en/sql-reference/table-functions/hdfs.md?plain=1#L125) but I cannot find it in the codebase, despite it being added in this PR. Was it removed? (couldn't find a PR which removed it)

**Comment by zvonand** [comment]
@Blargian removed in follow-up PR: https://github.com/ClickHouse/ClickHouse/pull/54863
